connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala

```diff
@@ -205,14 +220,26 @@ case class SessionHolder(userId: String, sessionId: String, session: SparkSessio
    * Called only by SparkConnectSessionManager.
    */
   private[connect] def close(): Unit = {
+    // It is called only by SparkConnectSessionManager.closeSession, only after it's removed from
+    // the sessionStore there guarantees that it is called only once.
+    if (closedTime.isDefined) {
+      throw new IllegalStateException(s"Session $key is already closed.")
+    }
```

Comment:
I was wondering if it needs extra synchronization, but concluded that not.
It is already explained below why it doesn't need extra synchronization against addExecuteHolder.